firedrake/dmplex.pyx

```diff
@@ -257,118 +260,149 @@ def closure_ordering(PETSc.DM plex,
+        PetscInt fStart, fEnd, vStart, vEnd
+        PetscInt entity_per_cell, ncells
+        PetscInt nclosure, p, vi, v, fi, i
+        PetscInt start_v, off
+        PetscInt *closure = NULL
+        PetscInt c_vertices[4]
+        PetscInt c_facets[4]
+        PetscInt g_vertices[4]
+        PetscInt vertices[4]
+        PetscInt facets[4]
+        int reverse
         np.ndarray[np.int32_t, ndim=2, mode="c"] cell_closure
-        np.ndarray[np.int32_t, ndim=2, mode="c"] edge_directions
 
     cStart, cEnd = plex.getHeightStratum(0)
-    ncells = cEnd - cStart
+    fStart, fEnd = plex.getHeightStratum(1)
+    vStart, vEnd = plex.getDepthStratum(0)
 
-    cell_closure = np.empty((ncells, nclosure), dtype=np.int32)
-    edge_directions = np.empty((ncells, e_per_cell), dtype=np.int32)
+    ncells = cEnd - cStart
+    entity_per_cell = 4 + 4 + 1
 
+    cell_closure = np.empty((ncells, entity_per_cell), dtype=np.int32)
     for c in range(cStart, cEnd):
         CHKERR(PetscSectionGetOffset(cell_numbering.sec, c, &cell))
-        cell_closure[cell, v_per_cell + e_per_cell] = c
-
-        CHKERR(DMPlexGetCone(plex.dm, c, &cone))
-        CHKERR(DMPlexGetConeOrientation(plex.dm, c, &cone_orient))
-
-        # We use two values only for edge orientation:
-        #  0: nodes are located on the edge in straight order
-        #  1: nodes are located on the edge in reverse order
-        for i in range(0, 4):
-            o = cone_orient[i]
-            edge_direction[i] = -(o+1) if o < 0 else o
+        CHKERR(DMPlexGetTransitiveClosure(plex.dm, c, PETSC_TRUE, &nclosure, &closure))
 
-        # DMPlex returns the edges in an order like this:
+        # First extract the facets (edges) and the vertices
+        # from the transitive closure into c_facets and c_vertices.
+        # Here we assume that DMPlex gives entities in the order:
         #
-        #   *--2--*
+        #   8--3--7
         #   |     |
-        #   3     1
+        #   4  0  2
         #   |     |
-        #   *--0--*
+        #   5--1--6
         #
-        # (Any edge can be the first edge, and the second edge
-        # can be any of its neighbours.)
+        # where the starting vertex and order of traversal is arbitrary.
+        # (We fix that later.)
         #
-        # The edges are directed like this:
+        # For the vertices, we also retrieve the global numbers into g_vertices.
+        vi = 0
+        fi = 0
+        for p in range(nclosure):
+            if vStart <= closure[2*p] < vEnd:
+                CHKERR(PetscSectionGetOffset(vertex_numbering.sec, closure[2*p], &v))
+                c_vertices[vi] = closure[2*p]
+                g_vertices[vi] = cabs(v)
+                vi += 1
+            elif fStart <= closure[2*p] < fEnd:
+                c_facets[fi] = closure[2*p]
+                fi += 1
+
+        # Cell orientation is the global number of the first vertex.
+        # The second vertex is always the one with the smaller global number.
```

Comment:
"so I find the smallest global numbered vertex and then I have a choice" => No, you find the vertex given in the cell orientations, that is the first vertex. Then you have a choice.
"you go to whichever has the smaller global number?" => Yes. The point is -- as described in a comment in orientations_facet2cell -- that for cells in the halo, the same choice must be made in all MPI processes, otherwise the interpretation of the cell interior dofs might be inconsistent among MPI processes.